rlberry/agents/utils/replay.py

```diff
@@ -0,0 +1,347 @@
+            for tag in self.tags:
+                self._data[tag][position] = data[tag]
+
+        #
+        # Priorities
+        #
+        if self._enable_prioritized:
+            self._it_sum[self._position] = self._max_priority**self._alpha
+            self._it_min[self._position] = self._max_priority**self._alpha
+
+        # Faster than append and pop
+        self._position = (self._position + 1) % self._max_replay_size
+
+    def end_episode(self):
+        """Call this method to indicate the end of an episode."""
+        self._current_episode += 1
+
+    def _sample_proportional_and_compute_weights(self, batch_size):
+        """Used for prioritized replay."""
+        # sample indices
+        indices = []
+        p_total = self._it_sum.sum(0, len(self) - 1)
+        every_range_len = p_total / batch_size
+        for i in range(batch_size):
+            mass = self._rng.uniform() * every_range_len + i * every_range_len
+            idx = self._it_sum.find_prefixsum_idx(mass)
+            indices.append(idx)
+        indices = np.array(indices)
+
+        # importance weights
+        weights = []
+        p_min = self._it_min.min() / self._it_sum.sum()
+        max_weight = (p_min * len(self)) ** (-self._beta)
+
+        for idx in indices:
+            p_sample = self._it_sum[idx] / self._it_sum.sum()
+            weight = (p_sample * len(self)) ** (-self._beta)
+            weights.append(weight / max_weight)
+        weights = np.array(weights)
+
+        return indices, weights
+
+    def _sample_batch_indices(self, batch_size, chunk_size, sampling_mode):
+        current_size = len(self)
+        if sampling_mode == "uniform":
+            start_indices = self._rng.choice(current_size - chunk_size, size=batch_size)
+            weights = np.ones_like(start_indices, dtype=np.float32)
+        elif sampling_mode == "prioritized":
+            start_indices, weights = self._sample_proportional_and_compute_weights(
+                batch_size=batch_size
+            )
+            start_indices[start_indices > (current_size - chunk_size)] = (
+                current_size - chunk_size
+            )
+            assert np.all(start_indices <= (current_size - chunk_size))
+        else:
+            raise ValueError(f"Invalid sampling_mode: '{sampling_mode}'")
+        end_indices = start_indices + chunk_size
+
+        # Handle invalid sub-trajectories: those containing a transition
+        # corresponding to self._position. That is, we need to handle
+        # the fact that we're sampling sub-trajectories from a circular buffer
+        # and that we cannot cross the 'boundary' defined by self._position.
+        #
+        # What we do here is to move the start_index to the most recent
+        # index (with respect to self._position) that allows sampling
+        # a sub-trajectory of length chunk_size.
+        # Note that this can result in a negative start_index,
+        # which is handled in the _get_chunk method.
```

Comment:
Indeed, that's the trickiest part of this implementation. I thought about 2., but it would make the runtime stochastic, which I don't think is desirable. We're indeed biasing the distribution, but the bias decreases as the size of the buffer increases (and no bias for chunk_size=1), so I think the option 1. (keep it like this) is ok.

I haven't given it much thought, but I couldn't immediately figure out a simple way of sampling over a distribution over chunks in O(batch_size * chunk_size) deterministic time.

I'd say (little bias + no rejection sampling) is better than (no bias + rejection sampling). What do you think?

Reply:
I think that if you could compute the chunk probabilities in the prioritized case by making a few adjustments to the `_sample_proportional_and_compute_weights` function:

1. Instead of setting `p_total = it_sum.sum(0, len(self) - 1)` sum only over the valid indices;
2. At each range, instead of using the sampled `mass` directly, call `find_prefixsum_idx` with the `mass` plus the mass of the indices we're ignoring.

This would look like something like this:

```python
# Valid indices are part of the range [start_idx, end_idx]
start_mass, end_mass = self._it_sum.sum(0, start_idx), self._it_sum.sum(0, end_idx)

# Computes the sum of priorities from start_idx to end_idx
if start_mass <= end_mass:
  p_total = self._it_sum.sum(0, len(self)-1) - end_mass + start_mass
else:
  p_total = start_mass - end_mass

# When computing the index from a sample, we must add the ignored mass again
mass = ...
if start_mass <= end_mass:
  mass += start_mass
elif mass > start_mass:
  mass += start_mass - end_mass
idx = self._it_sum.find_prefixsum_idx(mass)
```

Reply:
Of course, this method also introduces some bias because it sets the chunk priority to the priority of the first element of the chunk (one could argue that maybe choosing the last element is better). I think any of these implementations (including your original) is fine, but it would be interesting to test if these choices matter.

Reply:
I just pushed a commit to remove the bias from the `"uniform"` sampling mode. For prioritized replay, I think it's better to keep it like this for now: it's simpler and the bias only increases a little the probability of sampling the last chunk; and, in prioritized experience replay, the last transitions are initialized with the highest priority anyway. So I think the impact of this choice shouldn't matter too much.